sensor/common/image/service_impl_test.go

```diff
@@ -0,0 +1,280 @@
+			expectRegistry:   expectRegistryHelper(suite.mockRegistryStore, 1, true),
+			expectLocalScan:  expectLocalScan(suite.mockLocalScan, 1, nil, err),
+			expectedError:    errLocalScan,
+			expectedResponse: nil,
+		},
+	}
+	for testName, c := range cases {
+		suite.T().Run(testName, func(t *testing.T) {
+			suite.service.centralReady.Reset()
+			suite.service.Notify(c.notify)
+			c.expectCache.Fn()()
+			c.expectRegistry.Fn()()
+			c.expectCentralCall.Fn()()
+			c.expectLocalScan.Fn()()
+			res, err := suite.service.GetImage(ctx, c.request)
+			if c.expectedError != nil {
+				assert.EqualError(t, err, c.expectedError.Error())
+			} else {
+				assert.NoError(t, err)
+			}
+			assert.Equal(t, c.expectedResponse, res)
+		})
+	}
+}
+
+func expectCacheHelper(mockCache *cacheMocks.MockCache, times int, retValue interface{}) *expectFunctionHelper {
+	var fn func()
+	if times == 0 {
+		fn = func() {
+			mockCache.EXPECT().Get(gomock.Any()).Times(0)
+		}
+	} else {
+		fn = func() {
+			mockCache.EXPECT().Get(gomock.Any()).Times(times).DoAndReturn(func(_ interface{}) interface{} {
+				return retValue
+			})
+		}
+	}
+	return &expectFunctionHelper{
+		fn: fn,
+	}
+}
+
+func expectRegistryHelper(mockRegistryStore *imageMocks.MockregistryStore, times int, retValue bool) *expectFunctionHelper {
+	var fn func()
+	if times == 0 {
+		fn = func() {
+			mockRegistryStore.EXPECT().IsLocal(gomock.Any()).Times(0)
+		}
+	} else {
+		fn = func() {
+			mockRegistryStore.EXPECT().IsLocal(gomock.Any()).Times(times).DoAndReturn(func(_ interface{}) bool {
+				return retValue
+			})
+		}
+	}
+	return &expectFunctionHelper{
+		fn: fn,
+	}
+}
+
+func expectCentralCall(mockCentral *imageMocks.MockcentralClient, times int, retValue *v1.ScanImageInternalResponse, retErr error) *expectFunctionHelper {
+	var fn func()
+	if times == 0 {
+		fn = func() {
+			mockCentral.EXPECT().ScanImageInternal(gomock.Any(), gomock.Any()).Times(0)
+		}
+	} else {
+		fn = func() {
+			mockCentral.EXPECT().ScanImageInternal(gomock.Any(), gomock.Any()).Times(times).DoAndReturn(func(_, _ interface{}, _ ...interface{}) (*v1.ScanImageInternalResponse, error) {
+				return retValue, retErr
+			})
+		}
+	}
+	return &expectFunctionHelper{
+		fn: fn,
+	}
+}
+
+func expectLocalScan(mockLocalScan *imageMocks.MocklocalScan, times int, retValue *storage.Image, retErr error) *expectFunctionHelper {
+	var fn func()
+	if times == 0 {
+		fn = func() {
+			mockLocalScan.EXPECT().EnrichLocalImageInNamespace(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Times(0)
+		}
+	} else {
+		fn = func() {
+			mockLocalScan.EXPECT().EnrichLocalImageInNamespace(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Times(times).DoAndReturn(func(_, _, _, _, _, _ interface{}) (*storage.Image, error) {
+				return retValue, retErr
+			})
+		}
+	}
+	return &expectFunctionHelper{
+		fn: fn,
+	}
+}
+
+type expectFunctionHelper struct {
```

Comment:
Why use a struct with a single function pointer when you can return the function pointer itself? 

Wouldn't
```go
func expectX() func() {
      return func () { /* expect call */ } 
}
```

Reply:
The reason why I added the struct is to be able to leave some expectValues empty in the test definitions. 

```golang
func (e *expectFunctionHelper) Fn() func() {
	if e == nil || e.fn == nil {
		return func() {}
	}
	return e.fn
}
```

This creates an empty function (if expectFunctionHelper is nil) that can be called later without worrying about nil values:

```golang
c.expectCache.Fn()()
c.expectRegistry.Fn()()
c.expectCentralCall.Fn()()
c.expectLocalScan.Fn()()
```

Reply:
I see, but I still think there are more better ways to achieve that. If you want to keep the functions `nil`, you could use a slice of "expected" functions and just iterate through them and check nil on the loop. Or you could create an `expected` function that takes any number of functions as parameter, like [this](https://go.dev/play/p/Q8CXWdf39PL).

On the other hand, if you'd like to keep a `type` abstraction for these, you can still do it with a func type. This a much more esoteric approach. I've explored it [here](https://go.dev/play/p/ctm51sNUkg0). 


The overhead of having to deal with this struct in every `expect*` function is what ultimately bothers me in this approach.

Reply:
I like the esoteric approach 🙂 I've changed it to use the `type` abstraction with a func. Let me know what you think.